src/main/kotlin/no/nav/tilleggsstonader/sak/vedtak/læremidler/beregning/LæremidlerBeregningService.kt

```diff
@@ -1,49 +1,132 @@
+import no.nav.tilleggsstonader.sak.vedtak.læremidler.domain.Vedtaksperiode
+import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.domain.StønadsperiodeRepository
+import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.domain.MålgruppeType
+import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.domain.ResultatVilkårperiode
+import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.domain.VilkårperiodeRepository
 import org.springframework.stereotype.Service
 import java.math.BigDecimal
 import java.math.RoundingMode
-import java.time.YearMonth
 
 private val PROSENT_50 = BigDecimal(0.5)
 private val PROSENTGRENSE_HALV_SATS = 50
 
 @Service
-class LæremidlerBeregningService {
+class LæremidlerBeregningService(
+    private val vilkårperiodeRepository: VilkårperiodeRepository,
+    private val stønadsperiodeRepository: StønadsperiodeRepository,
+) {
     /**
      * Beregning av læremidler er foreløpig kun basert på antakelser.
      * Nå beregnes det med hele måneder, splittet i månedsskifte, men dette er ikke avklart som korrekt.
      * Det er ikke tatt hensyn til begrensninger på semester og maks antall måneder i et år som skal dekkes.
      */
 
-    fun beregn(periode: BeregningPeriode): BeregningsresultatLæremidler {
-        val beregningsgrunnlagPerMåned = periode.splitPerMåned { måned, periode ->
-            val beregningsGrunnlag = lagBeregningsGrunnlag(periode, måned)
-            BeregningsresultatForMåned(
-                beløp = finnBeløpForStudieprosent(beregningsGrunnlag.sats, beregningsGrunnlag.studieprosent),
-                grunnlag = lagBeregningsGrunnlag(periode, måned),
+    // ANTAR: En aktitivet hele vedtaksperioden
+    // ANTAR: En stønadsperiode per vedtaksperiode
+    // ANTAR: Sats ikke endrer seg i perioden
+    fun beregn(vedtaksperioder: List<Vedtaksperiode>, behandlingId: BehandlingId): BeregningsresultatLæremidler {
+        val stønadsperioder =
+            stønadsperiodeRepository.findAllByBehandlingId(behandlingId).tilSortertGrunnlagStønadsperiode()
+
+        validerVedtaksperioder(vedtaksperioder, stønadsperioder)
+
+        val aktiviteter = finnAktiviteter(behandlingId)
+
+        val beregningsresultatForMåned = vedtaksperioder.flatMap { vedtaksperiode ->
+            val utbetalingsperioder = vedtaksperiode.delTilUtbetalingsPerioder()
+
+            utbetalingsperioder.map { utbetalingsperiode ->
+                val relevantStønadsperiode = utbetalingsperiode.finnRelevantStønadsperiode(stønadsperioder)
+
+                lagBeregningsresultatForMåned(
+                    utbetalingsperiode = utbetalingsperiode,
+                    stønadsperiode = relevantStønadsperiode,
+                    aktivitet = utbetalingsperiode.finnRelevantAktivitet(
+                        aktiviteter = aktiviteter,
+                        aktivitetType = relevantStønadsperiode.aktivitet,
+                    ),
+                )
+            }
+        }
```

Comment:
Denne delen kan vurderes å trekkes ut i en metode. Då får man en ren metode med det som kun gjelder beregning, uten uthenting av data og validering.
```kotlin
val perioder = beregnLæremidlerPerMåned(vedtaksperioder, stønadsperioder, aktiviteter)
return BeregningsresultatLæremidler(perioder)
```

For å unngå map inne i en map så kan man flate ut det først
```suggestion
        val beregningsresultatForMåned = vedtaksperioder
            .flatMap { it.delTilUtbetalingsPerioder() }
            .map { utbetalingsperiode ->
                val relevantStønadsperiode = utbetalingsperiode.finnRelevantStønadsperiode(stønadsperioder)

                lagBeregningsresultatForMåned(
                    utbetalingsperiode = utbetalingsperiode,
                    stønadsperiode = relevantStønadsperiode,
                    aktivitet = utbetalingsperiode.finnRelevantAktivitet(
                        aktiviteter = aktiviteter,
                        aktivitetType = relevantStønadsperiode.aktivitet,
                    ),
                )
            }
            return BeregningsresultatLæremidler(beregningsresultatForMåned)
```

Reply:
Det ble ryddig 👍  026e151b48fa44da22e3a0997a41c152720c72f9